src/DynamoCoreWpf/Views/Search/LibraryView.xaml.cs

```diff
@@ -146,10 +146,56 @@ private void OnTreeViewItemPreviewMouseLeftButton(object sender, MouseButtonEven
             {
                 if (wrapPanel.MakeOrClearSelection(selectedClass))
                 {
-                    e.Handled = true;
+                    // If class button was clicked, then handle, otherwise leave it.
+                    e.Handled = selectedClass.SubCategories.Count == 0;
                     selectedElement.BringIntoView();
                 }
             }
+
+            ExpandCategory(categoryButton, selectedClass);
+            e.Handled = !(selectedClass is RootNodeCategoryViewModel);
+        }
+
+        private void ExpandCategory(TreeViewItem sender, NodeCategoryViewModel selectedClass)
+        {
+            // Get all category items.
+            var categories = sender.Items.OfType<NodeCategoryViewModel>();
+
+            // Get all current expanded categories.
+            List<NodeCategoryViewModel> allExpandedCategories = categories.
+                Where(cat => (!(cat is ClassesNodeCategoryViewModel) && cat.IsExpanded == true)).ToList();
+
+            var categoryToBeExpanded = categories.Where(cat => cat == selectedClass).FirstOrDefault();
+
+            // If categoryToBeExpanded is null, that means not category button, but class button was clicked.
+            // During loop we will find out to which category this clicked class belongs.
+            if (categoryToBeExpanded != null)
+                categoryToBeExpanded.IsExpanded = !categoryToBeExpanded.IsExpanded;
+
+            // Get expanded categories that should be collapsed.
+            var categoriesToBeCollapsed = allExpandedCategories.Remove(categoryToBeExpanded);
+
+            // Close all open categories, except one, that contains class.
+            // Or if category was clicked, also expand it and close others.
+            foreach (var categoryToBeCollapsed in allExpandedCategories)
+            {
+                // If class button was clicked.
+                if (categoryToBeExpanded == null)
+                {
+                    var categoryClasses = categoryToBeCollapsed.Items[0] as ClassesNodeCategoryViewModel;
+                    // Ensure, that this class is not part of current category.
+                    if (categoryClasses != null)
+                        if (categoryClasses.Items.Contains(selectedClass))
+                            categoryToBeCollapsed.IsExpanded = true;
+                        else
+                            categoryToBeCollapsed.IsExpanded = false;
```

Comment:
This could have been:

```csharp
if (categoryClasses != null)
    categoryToBeCollapsed.IsExpanded = categoryClasses.Items.Contains(selectedClass);
```

Also, remember that `if` statements with multiple-line body need to have curly brackets :)